src/Dom/Document.php

```diff
@@ -635,22 +642,22 @@ public function normalizeDomStructure()
             }
 
             $this->body = $body;
-            $html->appendChild($this->body);
+            $html->appendChild($body);
 
-            if ($oldDocumentElement !== $this->body && $oldDocumentElement !== $this->head) {
-                $this->body->appendChild($oldDocumentElement);
+            if ($oldDocumentElement !== $body && $oldDocumentElement !== $this->head) {
+                $body->appendChild($oldDocumentElement);
             }
         } else {
             $head = $this->getElementsByTagName(Tag::HEAD)->item(0);
             if (!$head) {
-                $this->head = $this->createElement(Tag::HEAD);
-                $this->documentElement->insertBefore($this->head, $this->documentElement->firstChild);
+                $this->properties['head'] = $this->createElement(Tag::HEAD);
```

Comment:
Why not continue to use `$this->head` and `$this->body`? Is there a reason to switch to using `$this->properties`?

Reply:
Just for the sake of performance?

Reply:
I ran into some issues when I originally tried that, but I might have been doing something wrong. I can give it another try 👍

Reply:
Just tested again. Even on PHP 8.1 I am getting lots of `Undefined property` warnings in this case.

Even when doing something seemingly simple like this:

```php
// Throws warning
$this->html = $html;
return $this->html;

// Works
$this->html = $html;
return $html;

// Also works
$this->properties['html'] = $html;
return $this->properties['html'];
```

I also just analyzed the `AmpProject\Dom\DocumentTest::testNormalizeDomStructure()` test case for this.

1. Calling `normalizeDomStructure()` calls `$this->head`
2. This triggers `__get()`
3. The getter for `head` runs, and it sees that it isn't set yet, which causes a call to `normalizeDomStructure()` _again_. The property hasn't been set yet.
4. That one then calls `moveInvalidHeadNodesToBody` at the end
5. That one then triggers `Undefined property: AmpProject\Dom\Document::$head`.

No such issues happen with the current `$properties` approach.

Reply:
I think the issue at question is this [note](https://www.php.net/manual/en/language.oop5.overloading.php#object.get:~:text=PHP%20will%20not%20call%20an%20overloaded%20method%20from%20within%20the%20same%20overloaded%20method.):

> PHP will not call an overloaded method from within the same overloaded method. That means, for example, writing return $this->foo inside of [__get()](https://www.php.net/manual/en/language.oop5.overloading.php#object.get) will return null and raise an E_WARNING if there is no foo property defined, rather than calling [__get()](https://www.php.net/manual/en/language.oop5.overloading.php#object.get) a second time. However, overload methods may invoke other overload methods implicitly (such as [__set()](https://www.php.net/manual/en/language.oop5.overloading.php#object.set) triggering [__get()](https://www.php.net/manual/en/language.oop5.overloading.php#object.get)).

Reply:
In other words, if a caller is trying to access `$x->foo` which invokes a magic getter, the getter must not also try to access `$x->foo` although it can access `$x->bar`. So there can't be any recursive reference to the current magic property being gotten.